### PR TITLE
refactor(tcp_server): release state lock before sending packets

### PR DESCRIPTION
**Pull request by kurnevsky**
Currently we send tcp packets while holding the state lock. At the
same time we use bounded queues for sending of packets. This means
that a single stuck client can block the whole server until the
timeout happens. This change makes the server release the state lock
before sending any packet.

### CODE REVIEW

File: tox_core/src/relay/server/server.rs [hunk 1]
```diff
@@ -42,6 +46,54 @@ struct ServerState {
...
+    while futures.next().await.is_some() { };
+}
+
+enum HandleAction {
+    Send(Vec<(Packet, mpsc::Sender<Packet>)>),
+    Onion(OnionRequest, SocketAddr),
+}
+
+impl Default for HandleAction {
+    fn default() -> Self {
```
Comment (kpp) [inline tox_core/src/relay/server/server.rs:72]: impl HandleAction::ignore == HandleAction::default and use it instead.

File: tox_core/src/relay/server/server.rs [hunk 2]
```diff
@@ -253,7 +328,7 @@ impl Server {
                 // haven't received DisconnectNotification yet and have sent yet another packet.
                 // In this case we don't want to throw an error and force disconnect the second client.
                 // TODO: failure can be used to return an error and handle it inside ServerProcessor
-                return Ok(())
+                return Ok(HandleAction::default())
```
Comment (kpp) [inline tox_core/src/relay/server/server.rs:331]: HandleAction::ignore

**Review by kpp — CHANGES_REQUESTED** [review]
So basically this is release state lock + some refactoring. Looks good.

File: tox_core/src/relay/server/server.rs (hunk 1 above)
Comment (kurnevsky) [reply tox_core/src/relay/server/server.rs:72]: I'd call it `empty` then :)

**Review by kpp — APPROVED** [review]

**Pull request merged**